app/assets/javascripts/models/beta/iteration.js

```diff
@@ -156,52 +190,65 @@ const pastAverageVelocity = (pastIterations) => {
 }
 
 const addToSprintFromBacklog = (sprints, project, story, velocity) => {
-  const sprintIndex = sprints.length && sprints.length - 1;
-  const hasSpace = canTakeStory(sprints[sprintIndex], Story.getPoints(story));
-
-  if (hasSpace) {
-    return addStoryToSprint(project, sprints, sprintIndex, story, velocity);
-  }
-
   const iterationNumber = getNextIterationNumber(sprints);
 
-  sprints[sprintIndex + 1] = createSprint(
+  return [...sprints, createSprint(
     iterationNumber,
     getDateForIterationNumber(iterationNumber, project),
-    undefined,
-    velocity
-  );
-
-  return addStoryToSprint(project, sprints, sprintIndex + 1, story, velocity);
+    false,
+    velocity,
+    [story],
+    Story.getPoints(story)
+  )];
 };
 
 export const groupBySprints = (stories = [], project, initialSprintNumber = 1, pastIterations) => {
   const velocity = sprintVelocity(project, pastIterations)
 
   return stories.reduce((sprints, story) => {
     const firstSprintIndex = 0;
-    const isFromSprintInProgress = !Story.isUnstarted(story);
-
-    sprints = handleSprintsOverflow(
-      project,
-      sprints,
-      Story.getPoints(story),
-      initialSprintNumber,
-      velocity
-    );
+    const isFromSprintInProgress = Story.isInProgress(story);
+
+    if (isEmpty(sprints)) {
+      sprints = [createFirstSprint(project, initialSprintNumber, velocity)];
+    }
 
-    if (sprints.length === 0) {
-      createFirstSprint(sprints, project, initialSprintNumber, velocity);
+    if (causesOverflow(story, velocity) && !isFromSprintInProgress) {
+      return addSprintWithOverflow(project, sprints, story, velocity, initialSprintNumber);
     }
 
     if (isFromSprintInProgress) {
       return addStoryToSprint(project, sprints, firstSprintIndex, story, velocity);
     }
 
+    if (canEnterLastAddedSprint(sprints, story)) {
+      return addStoryToSprint(project, sprints, sprints.length - 1, story, velocity);
+    }
+
     return addToSprintFromBacklog(sprints, project, story, velocity);
   }, []);
 };
 
+const isFirstSprint = (sprint, initialSprintNumber) =>
+  sprint.number === initialSprintNumber;
+
+const isEmpty = array => !Boolean(array) || array.length === 0;
```

Comment:
Is this function really required? Since its only being called at the `stories.reduce()` with `[ ]` as initial value the `!Boolean(array)` part will always return false, therefore making it useless

Reply:
Oh, you are right. I will remove the `!Boolean(array)`. Tks!